creator-node/src/middlewares.js

```diff
@@ -144,63 +155,148 @@ async function getOwnEndpoint (req) {
+    // Initial delay before polling
     await utils.timeout(1000)
+
+    let discprovBlockNumber = -1
     for (let retry = 1; retry <= MaxRetries; retry++) {
       req.logger.info(`getCreatorNodeEndpoints retry #${retry}/${MaxRetries} || time from start: ${Date.now() - start2} discprovBlockNumber ${discprovBlockNumber} || blockNumber ${blockNumber}`)
+
       try {
         const fetchedUser = await libs.User.getUsers(1, 0, null, wallet)
+
         if (!fetchedUser || fetchedUser.length === 0 || !fetchedUser[0].hasOwnProperty('blocknumber') || !fetchedUser[0].hasOwnProperty('track_blocknumber')) {
           throw new Error('Missing or malformatted user fetched from discprov.')
         }
+
         user = fetchedUser
         discprovBlockNumber = Math.max(user[0].blocknumber, user[0].track_blocknumber)
+
         if (discprovBlockNumber >= blockNumber) {
           break
         }
       } catch (e) { // Ignore all errors until MaxRetries exceeded.
         req.logger.info(e)
       }
+
       await utils.timeout(RetryTimeout)
       req.logger.info(`getCreatorNodeEndpoints AFTER TIMEOUT retry #${retry}/${MaxRetries} || time from start: ${Date.now() - start2} discprovBlockNumber ${discprovBlockNumber} || blockNumber ${blockNumber}`)
     }
 
+    // Error if discprov doesn't return any user for wallet
+    if (!user) {
+      throw new Error(`Failed to retrieve user from discprov after ${MaxRetries} retries. Aborting.`)
+    }
+
+    // Error if discprov has still not indexed to target blockNumber
     if (discprovBlockNumber < blockNumber) {
       throw new Error(`Discprov still outdated after ${MaxRetries}. Discprov blocknumber ${discprovBlockNumber} requested blocknumber ${blockNumber}`)
     }
+  } else if (ensurePrimary && myCnodeEndpoint) {
+    /**
+     * Else if ensurePrimary required, polls discprov until it has indexed myCnodeEndpoint (for up to 60 seconds)
+     * Errors if retrieved primary does not match myCnodeEndpoint
+     */
+    req.logger.info(`getCreatorNodeEndpoints || no blockNumber passed, retrying until DN returns same endpoint`)
+
+    const start2 = Date.now()
+
+    // Will poll every 5 sec for up to 1 minute (60 sec)
+    const MaxRetries = 12
+    const RetryTimeout = 5000 // 5 seconds
+
+    // Initial delay before polling
+    await utils.timeout(1000)
+
+    let returnedPrimaryEndpoint = null
+    for (let retry = 1; retry <= MaxRetries; retry++) {
+      req.logger.info(`getCreatorNodeEndpoints retry #${retry}/${MaxRetries} || time from start: ${Date.now() - start2} myCnodeEndpoint ${myCnodeEndpoint}`)
+
+      try {
+        const fetchedUser = await libs.User.getUsers(1, 0, null, wallet)
+
+        if (!fetchedUser || fetchedUser.length === 0 || !fetchedUser[0].hasOwnProperty('creator_node_endpoint')) {
+          throw new Error('Missing or malformatted user fetched from discprov.')
+        }
+
+        user = fetchedUser
+        returnedPrimaryEndpoint = (user[0].creator_node_endpoint).split(',')[0]
```

Comment:
I think we should be using the libs fn to get the primary off the creator_node_endpoint